Multi-site SDA

para2x opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
In the multi-site SDA I try to use the modules/functions I developed for the re-factored SDA to perform state data assimilation for multiple sites at the same time. This function and the re-factored SDA share the same ensemble writing/running/reading, analysis, and adjustment functions, but the multi-site version internally reformats/reshapes the X, mu.f, Pf, Y and R so they could be used by analysis functions. 

There are two new functions (`Contruct.Pf`, `Construct.R`) developed and called specifically for multi-site which they mainly handle the reshaping of those variables. 

This PR I guess needs to wait for PR #2045 and PR #2042 to go through first, but I was hoping if I could get feedbacks on these two functions (`Contruct.Pf`, `Construct.R`). 
The multi-site SDA was tested and it works perfectly assuming the the PR #2045 review suggests that the re-factored SDA produces meaningful results.

<!--- Please select appropriate Priority, Status,and Type labels-->
<!--- If you do not have permission to select labels please state which labels you would like -->

## Description
<!--- Describe your changes in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## Review Time Estimate
<!---When do you want your code reviewed by?-->
- [ ] Immediately
- [ ] Within one week
- [ ] When possible
## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue) <!-- please add issue number -->
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the CHANGELOG.md.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

<!--this template is from https://www.talater.com/open-source-templates/#/page/99-->

tonygardella commented:
@para2x Put some of these cool graphics and explanation of sda [here](https://github.com/PecanProject/pecan/blob/develop/book_source/03_intermediate_users_guide/03_hidden_analyses/02_sda.Rmd)

para2x commented:
@mdietze this works all fine for me now; whenever PR #2045 gets merged I can bring my recent improvements in PR #2045 (as a result of addressing the comments) into this PR.

para2x commented:
@mdietze, 
I think, this PR is now ready for the first round of revisions.

mdietze commented on modules/assim.sequential/R/Multi_Site_Constructors.R:
```diff
@@ -0,0 +1,258 @@
+  nsite <- length(site.ids)
+  nvariable <- length(var.names)
+  # I will make a big cov matrix and then I will populate it with the cov of each site
+  pf.matrix <-matrix(0,(nsite*nvariable),(nsite*nvariable))
```
high level question: I don't understand why pf.matrix isn't just cov(X) on the full X. What am I missing in why you put so much work into calculating the matrix block-by-block

para2x replied:
This is the piece I was trying to explain today that, I do this do zero out the off-diagonal cov matrix between two sites.

mdietze commented on modules/assim.sequential/R/Multi_Site_Constructors.R:
```diff
@@ -0,0 +1,258 @@
+##' 
+##' @param x Vector of numbers to identify each block.
+##' @param b Numeric value for the size of the blocks within the matrix ordered depending on byrow
+##' @param byrow ogical value. If FALSE (the default) the blocks are filled by columns, otherwise the blocks in the matrix are filled by rows.
```
typo on logical

mdietze commented on modules/assim.sequential/R/Multi_Site_Constructors.R:
```diff
@@ -0,0 +1,258 @@
+  nsite <- length(site.ids)
+  nsite.ids.with.data <-length(site.ids.with.data)
+  nvariable <- length(var.names)
+  H <- matrix(0, (nobs * nsite.ids.with.data), (nvariable*nsite))
```
This process of setting nobs to the maximum number of observations seen at any one site seems like it could result in H having rows that are all zeros, which doesn't seem desirable. Why not just sum over the number of observations?

mdietze commented on modules/assim.sequential/R/sda.enkf_MultiSite.R:
```diff
@@ -0,0 +1,458 @@
+#' @import nimble
+#' @export
+#' 
+sda.enkf.multisite <- function(settings, obs.mean, obs.cov, Q = NULL, restart=F, 
```
This function looks largely identical to the existing SDA function, which both seems redundant and makes it really hard to see what's different between the two. Is there a reason to not just merge back into having just one SDA function?

para2x replied:
You're absolutely right. I need to make sure that everyone is done with their changes in SDA at some point and merge the two into one and basically make the re-factored the n=1 case of multi-site.

mdietze commented on modules/assim.sequential/R/sda.enkf_refactored.R:
```diff
@@ -209,6 +210,7 @@ sda.enkf <- function(settings, obs.mean, obs.cov, Q = NULL, restart=F,
     #------------------------------------------- Reading the output
     X_tmp <- vector("list", 2) 
     X <- list()
+    if (control$debug) browser()
```
should probably remove this debugging line before final PR merge

para2x replied:
I'm letting the user here to stop the code at different stages of the run and inspect the internal environment of the SDA function and when they are done with that, they can click on resume and the code will go on like it was never paused. This option can be set by setting `control$debug` to TRUE in sda argument.

para2x replied:
@araiho This could be an alternative for you instead of running the function line by line.

para2x commented:
I'll revert this, I should have pulled from develop first !

para2x commented on book_source/03_intermediate_users_guide/03_hidden_analyses/02_sda.Rmd:
```diff
@@ -199,3 +199,192 @@ $\boldsymbol{V}_{t+1}	=	n_{t}\bar{\boldsymbol{Q}}_{t}$
+    </format_id>
+    <input.id>
+      <input.id>1000013298</input.id>
+    </input.id>
```
I really don't use any of those tags and honestly I don't even know what is it used for.

mdietze replied:
1. If you don't use those tags, could you put them as I've requested
2. If you don't use them, how do you specify the data that needs to be read for the SDA?

mdietze commented on modules/assim.sequential/R/Analysis_sda.R:
```diff
@@ -277,20 +273,27 @@ GEF<-function(setting,Forecast,Observed, H, extraArg, nitr=50000, nburnin=10000,
   y.ind <- as.numeric(Y > interval[,1])
   y.censored <- as.numeric(ifelse(Y > interval[,1], Y, 0))
-  
+  browser()
```
This seems to be a debugging leftover. If it's meant to be there, it seems like it would be good to put in within an optional flag, e.g.
```
if(debug){
    browser()
}
```

para2x replied:
I still use arguments to pass the data to the sda function.

mdietze replied:
I don't understand -- your response about passing data to the function as an argument doesn't seem to have anything to do with the use of `browser()`

mdietze commented on modules/assim.sequential/inst/MultiSite-Exs/SDA/pecan.SDA.4sites.xml:
```diff
@@ -0,0 +1,202 @@
+    </format_id>
+    <input.id>
+      <input.id>1000013298</input.id>
+    </input.id>
```
FYI,  you changed this in the documentation, but not in this example xml

para2x commented:
hey @infotroph, just wanted to have your thoughts on something. 
Seems like the codes inside this PR is outdated and Github doesn't update them no matter how many times I push more changes to this branch. For example on Analysis_sda.R line 276 is deleted in latests commits (https://github.com/para2x/pecan/blob/MultiSite_SDA/modules/assim.sequential/R/Analysis_sda.R) but it's still here. Any idea what is going on ?
